metals/src/main/scala/scala/meta/internal/metals/watcher/SimpleFileWatcher.scala

```diff
@@ -0,0 +1,66 @@
+package scala.meta.internal.metals.watcher
+
+import collection.JavaConverters._
+import java.nio.file.{FileSystems, Path, StandardWatchEventKinds}
+import java.util.concurrent.BlockingQueue
+import scala.concurrent.{ExecutionContext, Future}
+
+object SimpleFileWatcher {
+  /*
+   * returns a function which closes the file watcher when called
+   */
+  def initWatcher(
+    watchFilter: Path => Boolean,
+    watchEventQueue: BlockingQueue[FileWatcherEvent],
+    filesToWatch: Set[Path],
+    directoriesToWatch: Set[Path],
+  )(implicit ec: ExecutionContext): () => Unit = {
+    scribe.error(s"RFERGUSON INIT WATCHER")
+    val watchService = FileSystems.getDefault().newWatchService();
+    filesToWatch.filter(watchFilter).foreach { path =>
+      scribe.info(s"watching parent directory of file $path")
+      // watchservice only watches directories
+      // this technically could watch things that weren't in the build, but shared a directory, but what can you do?
+      path.getParent().register(
+        watchService,
+        StandardWatchEventKinds.ENTRY_CREATE,
+        StandardWatchEventKinds.ENTRY_DELETE,
+        StandardWatchEventKinds.ENTRY_MODIFY,
+        //StandardWatchEventKinds.OVERFLOW,
+      )
+    }
+    directoriesToWatch.filter(watchFilter).foreach { path =>
+      // we don't really have to watch the parent directory, but then we don't have to have special logic for these down below
+      scribe.info(s"watching parent directory of directory $path")
+      path.getParent().register(
+        watchService,
+        StandardWatchEventKinds.ENTRY_CREATE,
+        StandardWatchEventKinds.ENTRY_DELETE,
+        StandardWatchEventKinds.ENTRY_MODIFY,
+      )
+    }
+    Future {
+      while (true) {
+        val key = watchService.take()
```

Comment:
Nit: You can clean this up like so:
```scala
for {
  key <- Option(watchService.take()).toList
  _ = key.reset()
  directory = key.watchable().asInstanceOf[Path]
  event <- key.pollEvents().asScala
  pathFromEventContext = event.context().asInstanceOf[Path]
  path = directory.resolve(pathFromEventContext)
  _ <- if (watchFilter(path)) List(()) else List.empty
} {
  scribe.info(s"got ${event.kind()} event for path: ${path}")

  watchEventQueue.add(
    event.kind() match {
      case StandardWatchEventKinds.ENTRY_CREATE | StandardWatchEventKinds.ENTRY_MODIFY =>
        FileWatcherEvent.createOrModify(path)

      case StandardWatchEventKinds.ENTRY_DELETE =>
        FileWatcherEvent.delete(path)
    }
  )
}
```

Reply:
So, I don't think I _can_ do the for comprehension, because `key.reset()` has to be called after I've called `pollEvents()`, but even if I could, I don't know that this is cleaner than what I have. There's like one/two actual monads in there, the rest is just what I have, but in a for comprehension.

Reply:
Ah, I didn't realize that.

> There's like one/two actual monads in there, the rest is just what I have, but in a for comprehension.

Fair enough. I just thought it'd be nice to minimize the indentation.